Shippable returns this error:
`````
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1537015424.75-158201857908434/AnsiballZ_docker_swarm.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1537015424.75-158201857908434/AnsiballZ_docker_swarm.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1537015424.75-158201857908434/AnsiballZ_docker_swarm.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 525, in <module>
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 520, in main
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 306, in __call__
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 329, in init_swarm
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 389, in __update_swarm
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 375, in __update_spec
KeyError: 'Labels'
`````
on the `[docker_swarm_service : Create a Swarm cluster]` task, which creates
a swarm using the `docker_swarm` module.  

Locally, on Ubuntu 17.10 and Centos7, I run both the `docker_swarm_service`
and the `docker_swarm` tests successfully.

I tried running the tests for `docker_swarm` module on a clean centos7 vm
with the distribution version of docker (`docker-1.13.1-74.git6e3bb8e.el7.centos.x86_64`)
and got the same `KeyError: 'Labels'` error.  
After upgrading docker as per https://docs.docker.com/install/linux/docker-ce/centos/
the was no more errors.

The `docker_swarm` fails when trying to read the `.Spec.Labels` value from a
dict returned from `client.inspect()`, which in turn executes a `GET` to
`/swarm`.  
I checked the responses from the docker apis on both versions and it seems like
the call to `/swarm` on a fresh stack returns:
- no 'Labels' attribute on the older version
  `````
  # curl -s --unix-socket /var/run/docker.sock http:/swarm|jq .Spec
  {
    "Name": "default",
    "Labels": {},
    "Orchestration": {
      "TaskHistoryRetentionLimit": 5
    },
    "Raft": {
      "SnapshotInterval": 10000,
      "KeepOldSnapshots": 0,
      "LogEntriesForSlowFollowers": 500,
      "ElectionTick": 10,
      "HeartbeatTick": 1
    },
    "Dispatcher": {
      "HeartbeatPeriod": 5000000000
    },
    "CAConfig": {
      "NodeCertExpiry": 7776000000000000
    },
    "TaskDefaults": {},
    "EncryptionConfig": {
      "AutoLockManagers": false
    }
  }
  `````
- a 'Label' attribute set to `{}` on the current version
  `````
  # curl -s --unix-socket /var/run/docker.sock http:/swarm|jq .Spec
  {
    "Name": "default",
    "Labels": {},
    "Orchestration": {
      "TaskHistoryRetentionLimit": 5
    },
    "Raft": {
      "SnapshotInterval": 10000,
      "KeepOldSnapshots": 0,
      "LogEntriesForSlowFollowers": 500,
      "ElectionTick": 10,
      "HeartbeatTick": 1
    },
    "Dispatcher": {
      "HeartbeatPeriod": 5000000000
    },
    "CAConfig": {
      "NodeCertExpiry": 7776000000000000
    },
    "TaskDefaults": {},
    "EncryptionConfig": {
      "AutoLockManagers": false
    }
  }
  ````` 

Browsing https://docs.docker.com/engine/api/v1.32/#operation/SwarmInspect
(and `v1.31`) I confirmed that the Labels attribute was added
to the sample response on API version 1.32.

I'd write a patch for the `docker_swarm` module defaulting `client.inspect()['Spec']['Labels']`
to `{}` but shippable [returned no errors](https://app.shippable.com/github/ansible/ansible/runs/78067/summary) when it was last updated.  
Has the shippable environment changed since?
Can someone reproduce this error?  
I feel like I might be missing something...

PS: I noticed that the `setup_docker` fragment installs a recent version of docker but
does not start the service: am I wrong in guessing that it does this to have the
docker client available in the test container while relaying on the parent docker
server (sharing the docker socket?) to execute the subsequent test target tasks?
If so is it possible to control the version of the docker server?

--- a/lib/ansible/modules/cloud/docker/docker_swarm_service.py
+++ b/lib/ansible/modules/cloud/docker/docker_swarm_service.py
@@ -262,13 +262,11 @@ options:
     - Maps to docker service --update-max-failure-ratio
   update_order:
     required: false
-    default: stop-first
+    default: null
     description:
     - Specifies the order of operations when rolling out an updated task.
     - Maps to docker service --update-order
-    choices:
-    - stop-first
-    - start-first
+    - Requires docker api version >= 1.29
   user:
     required: false
     default: root
@@ -515,7 +513,7 @@ class DockerService(DockerBaseClass):
         self.update_failure_action = "continue"
         self.update_monitor = 5000000000
         self.update_max_failure_ratio = 0.00
-        self.update_order = "stop-first"
+        self.update_order = None
 
     def get_facts(self):
         return {
@@ -890,7 +888,9 @@ class DockerServiceManager():
         ds.update_failure_action = update_config_data['FailureAction']
         ds.update_monitor = update_config_data['Monitor']
         ds.update_max_failure_ratio = update_config_data['MaxFailureRatio']
-        ds.update_order = update_config_data['Order']
+
+        if 'Order' in update_config_data:
+            ds.update_order = update_config_data['Order']
 
         dns_config = task_template_data['ContainerSpec'].get('DNSConfig', None)
         if dns_config:
@@ -1027,6 +1027,7 @@ class DockerServiceManager():
             {'param': 'tty', 'attribute': 'tty', 'min_version': '1.25'},
             {'param': 'secrets', 'attribute': 'secrets', 'min_version': '1.25'},
             {'param': 'configs', 'attribute': 'configs', 'min_version': '1.30'},
+            {'param': 'update_order', 'attribute': 'update_order', 'min_version': '1.29'}]
         params = self.client.module.params
         empty_service = DockerService()
         for pv in parameters_versions:
@@ -1156,7 +1157,7 @@ def main():
         update_failure_action=dict(default='continue', choices=['continue', 'pause']),
         update_monitor=dict(default=5000000000, type='int'),
         update_max_failure_ratio=dict(default=0, type='float'),
-        update_order=dict(default='stop-first', choices=['stop-first', 'start-first']),
+        update_order=dict(default=None, type='string'),
         user=dict(default='root'))
     required_if = [
         ('state', 'present', ['image'])

--- a/lib/ansible/modules/cloud/docker/docker_swarm_service.py
+++ b/lib/ansible/modules/cloud/docker/docker_swarm_service.py
@@ -177,7 +177,7 @@ options:
     - List of dictionaries describing the service configs.
     - Every item must be a dictionary exposing the keys config_id, config_name, filename, uid (defaults to 0), gid (defaults to 0), mode (defaults to 0o444)
     - Maps docker service --config option.
-    default: []
+    default: null
   networks:
     required: false
     default: []
@@ -498,7 +498,7 @@ class DockerService(DockerBaseClass):
         self.mode = "replicated"
         self.user = "root"
         self.mounts = []
-        self.configs = []
+        self.configs = None
         self.secrets = []
         self.constraints = []
         self.networks = []
@@ -629,16 +629,18 @@ class DockerService(DockerBaseClass):
             service_m['target'] = param_m['target']
             s.mounts.append(service_m)
 
-        s.configs = []
-        for param_m in ap['configs']:
-            service_c = {}
-            service_c['config_id'] = param_m['config_id']
-            service_c['config_name'] = str(param_m['config_name'])
-            service_c['filename'] = param_m.get('filename', service_c['config_name'])
-            service_c['uid'] = int(param_m.get('uid', "0"))
-            service_c['gid'] = int(param_m.get('gid', "0"))
-            service_c['mode'] = param_m.get('mode', 0o444)
-            s.configs.append(service_c)
+        s.configs = None
+        if ap['configs']:
+            s.configs = []
+            for param_m in ap['configs']:
+                service_c = {}
+                service_c['config_id'] = param_m['config_id']
+                service_c['config_name'] = str(param_m['config_name'])
+                service_c['filename'] = param_m.get('filename', service_c['config_name'])
+                service_c['uid'] = int(param_m.get('uid', "0"))
+                service_c['gid'] = int(param_m.get('gid', "0"))
+                service_c['mode'] = param_m.get('mode', 0o444)
+                s.configs.append(service_c)
 
         s.secrets = []
         for param_m in ap['secrets']:
@@ -756,18 +758,21 @@ class DockerService(DockerBaseClass):
                             read_only=mount_config['readonly'])
             )
 
-        configs = []
-        for config_config in self.configs:
-            configs.append(
-                types.ConfigReference(
-                    config_id=config_config['config_id'],
-                    config_name=config_config['config_name'],
-                    filename=config_config.get('filename'),
-                    uid=config_config.get('uid'),
-                    gid=config_config.get('gid'),
-                    mode=config_config.get('mode')
+        configs = None
+        if self.configs:
+            configs = []
+            for config_config in self.configs:
+                configs.append(
+                    types.ConfigReference(
+                        config_id=config_config['config_id'],
+                        config_name=config_config['config_name'],
+                        filename=config_config.get('filename'),
+                        uid=config_config.get('uid'),
+                        gid=config_config.get('gid'),
+                        mode=config_config.get('mode')
+                    )
                 )
-            )
+
         secrets = []
         for secret_config in self.secrets:
             secrets.append(
@@ -1021,7 +1026,7 @@ class DockerServiceManager():
             {'param': 'hostname', 'attribute': 'hostname', 'min_version': '1.25'},
             {'param': 'tty', 'attribute': 'tty', 'min_version': '1.25'},
             {'param': 'secrets', 'attribute': 'secrets', 'min_version': '1.25'},
-            {'param': 'configs', 'attribute': 'configs', 'min_version': '1.30'}]
+            {'param': 'configs', 'attribute': 'configs', 'min_version': '1.30'},
         params = self.client.module.params
         empty_service = DockerService()
         for pv in parameters_versions:
@@ -1118,7 +1123,7 @@ def main():
         image=dict(type='str'),
         state=dict(default="present", choices=['present', 'absent']),
         mounts=dict(default=[], type='list'),
-        configs=dict(default=[], type='list'),
+        configs=dict(default=None, type='list'),
         secrets=dict(default=[], type='list'),
         networks=dict(default=[], type='list'),
         args=dict(default=[], type='list'),

--- a/test/integration/targets/docker_swarm_service/tasks/test_swarm_service.yml
+++ b/test/integration/targets/docker_swarm_service/tasks/test_swarm_service.yml
@@ -133,3 +133,8 @@
     that:
     - output is success
     - output is not changed
+
+- name: Remove a Swarm cluster
+  docker_swarm:
+    state: absent
+    force: true

